### ISSUE TITLE
[BUG]: Object reference not set to an instance of an object

### ISSUE THREAD

**Issue by aropb** [SciSharp/LLamaSharp#891]
### Description

I use KernelMemory 0.69.240727.1. 

The error occurs at the time of the call:
memory.AskAsync()

![Screenshot_1](https://github.com/user-attachments/assets/e563258c-371e-41b5-940c-676f6f86fc97)

### Reproduction Steps

The error occurs at the time of the call:

MemoryAnswer answer = await memory.AskAsync(question: question, filters: filters);

### Environment & Configuration

- Operating system: 
- .NET runtime version: 8.0.7
- LLamaSharp version: 0.15.0
- CUDA version (if you are using cuda backend): -
- CPU & GPU device: CPU

### Known Workarounds

-

**Comment by jwangga** [comment]
I encountered the same issue when running the sample code:  "Kernel Memory: Document Q&A" or "Kernel Memory: Save and Load" from the LLama.Examples project.

**Comment by tusharmevl** [comment]
@aropb @jwangga I am facing the same issue running the example '**Kernel Memory: Document Q&A**'. Did you find a fix for this? I am trying to implement a RAG system using this. Is there any other way to implement it apart from Kernel Memory?

**Comment by jwangga** [comment]
@tusharmevl  I have not found a fix for the Kernel Memory issue.  It seems that the integration with Semantic Kernel Memory works. You may try using that as an alternative if your system only needs to support Text.

**Comment by tusharmevl** [comment]
@jwangga Ok Thanks! Yes I need to support text only for now, will try that.

**Comment by GalactixGod** [comment]
Thanks @jwangga !!

I'm seeing that you can use Semantic Kernel Memory (SKM) as well. 

Doesn't appear that you can "chat" with SKM to discuss results unfortunately. Have you been able to figure out a way to "ask" questions of SKM?

**Comment by nicholusi2021** [comment]
I'm also having the same issue with the Kernel Memory: Document Q&A example.

**Comment by aropb** [comment]
I really need to fix the error. So far, I can only use such versions:

<PackageReference Include="Microsoft.KernelMemory.Core" Version="0.62.240605.1" />
<PackageReference Include="LLamaSharp" Version="0.13.0" />

**Comment by aropb** [comment]
I found the place where the error occurs.

llama_get_logits_ith suddenly return null.

        public Span<float> SafeLLamaContextHandle.GetLogitsIth(int i)
        {
            var model = ThrowIfDisposed();

            unsafe
            {
                var logits = llama_get_logits_ith(this, i);
                return new Span<float>(logits, model.VocabCount);
            }
        }

But I don't understand what to do next and how to fix the error. Apparently, null shouldn't be there. Can anyone help with this error?

**Comment by martindevans** [comment]
That's probably indicative of two bugs in LLamaSharp.

## Wrapper Error

The docs for `llama_get_logits_ith` (see [here](https://github.com/ggerganov/llama.cpp/blob/9c1ba557335c3cab46807e6478eb7d17a63361f1/include/llama.h#L879-L883)) say:

```
// Logits for the ith token. For positive indices, Equivalent to:
// llama_get_logits(ctx) + ctx->output_ids[i]*n_vocab
// Negative indicies can be used to access logits in reverse order, -1 is the last logit.
// returns NULL for invalid ids.
LLAMA_API float * llama_get_logits_ith(struct llama_context * ctx, int32_t i);```

So it _is_ valid for `llama_get_logits_ith` to return null! That means this `SafeLLamaContextHandle.GetLogitsIth` is incorrectly written, it should check for null and raise some kind of error in that case (throw an exception most likely). It is never valid to pass a null pointer into a span constructor!

This is why you get a hard crash instead of an exception.

## Higher Level Error

`llama_get_logits_ith` returns null if an invalid value for `i` is passed in. There must be a bug somewhere in the higher level that is causing an incorrect value to be passed in. Since this error only seems to affect kernel memory it must be something specific to the KM wrapper.

**Comment by aropb** [comment]
I have found a solution.

1. Embeddings = false
...
       public static IKernelMemoryBuilder WithLLamaSharp(this IKernelMemoryBuilder builder, LLamaSharpConfig config)
        {
            ModelParams parameters = new(config.ModelPath)
            {
                Embeddings = false,
...

2. set the values:

                UBatchSize = 2000,
                BatchSize = 2000

...

        public LLamaSharpTextEmbeddingGenerator(LLamaSharpConfig config, LLamaWeights weights)
        {
            ModelParams @params = new(config.ModelPath)
            {
                Embeddings = true,
...
                UBatchSize = 2000,
                BatchSize = 2000
            };

**Comment by aropb** [comment]
While testing, I noticed that it became slower to work, about 2 times after 0.13.0. Why is this interesting?

**Comment by martindevans** [comment]
> Embeddings = false

Aha, I think you've cracked it! A while ago the behaviour of the embeddings flag was changed, so logits can no longer be extracted if `embeddings=true`.

**Comment by aropb** [comment]
And in LLamaSharpTextEmbeddingGenerator must specify the values UBatchSize, BatchSize!

**Comment by martindevans** [comment]
I'm not sure about that - there should be sensible defaults for those values. In LLamaSharp they're set to default values [here](https://github.com/SciSharp/LLamaSharp/blob/master/LLama/Common/ModelParams.cs#L49-L52). It's possible KernelMemory is overriding those defaults with something incorrect though (I don't really know the KM stuff, so I can't be certain).

**Comment by aropb** [comment]
Without these values, there will be an error "Input contains more tokens than configured batch size". And now you can only define them by rewriting the LLamaSharpTextEmbeddingGenerator class.

**Comment by aropb** [comment]
Apparently it is necessary to add UBatchSize, BatchSize to LLamaSharpConfig

**Comment by martindevans** [comment]
I'm super busy this month, but I will try to make time to fix the issues you found that I summarised [here](https://github.com/SciSharp/LLamaSharp/issues/891#issuecomment-2324667660) when I get a chance (soon, hopefully. Definitely before the next release).

**Comment by martindevans** [comment]
#920 Fixes the lowest level wrapper error, so at least it throws an exception. Hopefully that might help debug the higher level issue.

**Comment by aropb** [comment]
The problem has been found. You need to force embeddings=false.

**Comment by martindevans** [comment]
I wasn't sure if there's more going on, since you also mentioned a need to change the batch size. Is that just because of the size of your request (you need a larger batch to fit it all in), or is there more going on there?

**Comment by aropb** [comment]
Yes, the block size is larger than batchSize, but now this value cannot be changed except to rewrite the class LLamaSharpTextEmbeddingGenerator.

**Comment by eocron** [comment]
Any update on this?

**Comment by aropb** [comment]
> Any update on this?

There is a solution above

**Comment by jwangga** [comment]
@aropb Where should "Embeddings = false" be added?  There does not seem to be the method WithLLamaSharp in LLamaSharp.KernelMemory project. Thanks.

**Comment by aropb** [comment]
It is indicated above where he is

**Comment by sangyuxiaowu** [comment]
It seems that for models that support ChatCompletion and Embeddings, the new version must configure Embeddings=false in order to use ChatCompletion properly.

**Comment by jwangga** [comment]
It does not work for me, either.  I made the suggested changes in these places: 
In BuilderExtensions.cs
![image](https://github.com/user-attachments/assets/a8769744-098d-4e47-8b36-81feba49b57d)
In LLamaSharpTextEmbeddingGenerator
![image](https://github.com/user-attachments/assets/ce5cdab6-83a4-4f74-b998-a128f3672db0)

Am I missing something?

**Comment by aropb** [comment]
You need to always set the default Embeddings = false

**Comment by lubotorok** [comment]
I did what is mentioned here, see:

https://github.com/lubotorok/LLamaSharp/commit/d38091d4d33fb3281c6f0ec5f6d562cab1be334c

but I had to lower the Context Size too. Currently I set it to 4000, I had 131 000 before and I was getting Access Violation with llama-3.1-8b-4k model even with this modification. 

I am using the same model as a chat assistant with context size 131 000 and it works. 
I am just learning both llamasharp and KM. I hope this observation somehow helps.
